DynamicTablesPkg/Library/Acpi/Common/AcpiDbg2Lib/Dbg2Generator.c

```diff
@@ -266,33 +164,160 @@ FreeDbg2TableEx (
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Invalid parameters to PopulateDbg2Device\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (MemoryRangesCount == 0) {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Memory ranges count cannot be zero\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Check string length before concatenation
+  if ((DeviceInfo->ObjectName[0] != '\0') &&
+      (AsciiStrLen (SB_SCOPE) + AsciiStrLen (DeviceInfo->ObjectName) >= DBG2_NAMESPACESTRING_FIELD_SIZE))
+  {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Namespace string too long\n"));
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  if (MemoryRangesCount > MAX_UINT8) {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Too many memory ranges. Count = %u\n", MemoryRangesCount));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Validate all memory ranges
+  for (Index = 0; Index < MemoryRangesCount; Index++) {
+    if (MemoryRanges[Index].BaseAddress == 0) {
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Memory range base address is 0. Index = %u\n", Index));
+      return EFI_INVALID_PARAMETER;
+    }
+
+    if (MemoryRanges[Index].Length == 0) {
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Memory range length is 0. Index = %u\n", Index));
+      return EFI_INVALID_PARAMETER;
+    }
+
+    if (MemoryRanges[Index].Length > MAX_UINT32) {
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Memory range length too large. Length = %u\n", MemoryRanges[Index].Length));
+      return EFI_INVALID_PARAMETER;
+    }
+  }
+
+  // Calculate total size with overflow check
+  TotalSize = (UINT64)sizeof (EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT) +
+              ((UINT64)sizeof (EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE) + sizeof (UINT32)) * MemoryRangesCount +
+              (UINT64)sizeof (CHAR8) * DBG2_NAMESPACESTRING_FIELD_SIZE;
+
+  if (TotalSize > MAX_UINT16) {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Device size too large\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Dbg2DeviceSize    = (UINT16)TotalSize;
+  CurrentDbg2Device = *AcpiDbg2Device;
+
+  // Initialize device structure
+  CurrentDbg2Device->Revision                        = EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT_REVISION;
+  CurrentDbg2Device->Length                          = Dbg2DeviceSize;
+  CurrentDbg2Device->NumberofGenericAddressRegisters = (UINT8)MemoryRangesCount;
+  CurrentDbg2Device->NameSpaceStringLength           = DBG2_NAMESPACESTRING_FIELD_SIZE;
+  CurrentDbg2Device->NameSpaceStringOffset           = sizeof (EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT);
+  CurrentDbg2Device->OemDataLength                   = 0;
+  CurrentDbg2Device->OemDataOffset                   = 0;
+  CurrentDbg2Device->PortType                        = DeviceInfo->PortType;
+  CurrentDbg2Device->PortSubtype                     = DeviceInfo->PortSubtype;
+  CurrentDbg2Device->BaseAddressRegisterOffset       = CurrentDbg2Device->NameSpaceStringOffset + DBG2_NAMESPACESTRING_FIELD_SIZE;

```

Comment:
similarly, the list of BaseAddressRegister fields should be before the namespace string